### PR TITLE
Add jobs signals to schedular portal

### PR DESCRIPTION
**Pull request by oanaBiancaSchiopu**
On the right click action for a job, there is a new menu (Actions) which contains a new sub-menu showing the job's signals
The Actions menu will be enabled if these 3 conditions are true:
- the job is running 
- there is just one job selected
- the job has signals

The selected signal will be send to the job
There are 2 request done 
- one that retrives the signals when right click on a job
- another one that send the signal to the job

Capture

### CODE REVIEW

**Comment by amouhoub** [comment]
@oanaBiancaSchiopu Can you please add one or more screenshot to the description of your PR as it introduces a new UI component that is hard to review visually based on code.

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/json/SchedulerJSONUtils.java
Comment (mbenguig) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/json/SchedulerJSONUtils.java:36]: You confirm importing client.* is not >> than importing the 5 ones?

**Review by mbenguig — APPROVED** [review]
A very short comment !

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/json/SchedulerJSONUtils.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/json/SchedulerJSONUtils.java:36]: I have used the optimize imports from intellij, for me it avoids long lists of imports. I am not sure it has a noticeable impact in memory consumption

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:493]: This code has to be in the controller not in the view

**Review by mboussaa — CHANGES_REQUESTED** [review]
To move the rest call from the view to the controller

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:493]: I did like this in the begining but I needed to get the signals before creating the Action menu and since the callback is asynchronous I had to create the menu when I receive the signals (on onSuccess method) so there would be no delay

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/json/SchedulerJSONUtils.java
Comment (AlexFalessi) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/json/SchedulerJSONUtils.java:284]: In my opinion Java8 style is simpler to read. 

For instance:

```
    public static Set<String> extractSet(JSONValue setValue) {
        if (setValue == null || setValue.isArray() == null) {
            return Collections.emptySet();
        } else {
            JSONArray keyValueArray = setValue.isArray();
            return IntStream.range(0, keyValueArray.size())
                        .mapToObj(keyValueArray::get)
                        .map(JSONValue::isString)
                        .map(JSONString::stringValue)
                        .collect(Collectors.toSet());
        }
    }
```

Your call

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (AlexFalessi) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:30]: Watch out with Intellij optimize imports. It can decide to import whole packages.
I guess it is not worth importing whole util here

Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:30]: I have added in plus these 2 packages:
import java.util.Set;
import java.util.HashSet;
The wildcard will replace 5 imports

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java
Comment (fviale) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java:456]: please add javadoc info

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerServiceAsync.java
Comment (fviale) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerServiceAsync.java:413]: please add javadoc info

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java
Comment (fviale) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java:798]: you can use `@Produces(MediaType.APPLICATION_JSON)`

**Review by fviale — APPROVED** [review]
LGTM, minor comments

**Comment by fviale** [comment]
Retest this please

**Comment by fviale** [comment]
@oanaBiancaSchiopu you need to run spotlessApply on your code

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:266]: sent

Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:266]: signal signal => signal the signal

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:273]: what does the result return? true/false? 
if it returns a boolean so we need to add info in the log message about the signal sent + jobid

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java
Comment (mboussaa) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java:456]: return Boolean ?

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:278]: same as above

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:524]: why critical ? 
What happens when a job has 0 actions ?

**Comment by fviale** [comment]
Retest this please

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:80]: comment to explain this new added attribute

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:524]: When the job has 0 actions, we will receive an empty list and the menu "Actions" will be disabled

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (mboussaa) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:493]: you can call a method from the controller + pass the needed widgets as parameters to the controller method

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java:799]: Set<String> ?

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/SchedulerServiceImpl.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/SchedulerServiceImpl.java:1372]: not void

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (mboussaa) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:524]: good, that what I was expecting. Thx

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java:458]: the current session id

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerService.java:459]: the signal to be sent to the job

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerServiceAsync.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerServiceAsync.java:415]: same

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerServiceAsync.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/SchedulerServiceAsync.java:416]: same

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (mboussaa) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java]: still waiting for an answer

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java
Comment (mboussaa) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java:799]: still waiting for an answer

**Comment by mboussaa** [comment]
Not very fan of the Action logo and @amouhoub as well agree. We suggest to remove the logo and keep it without logo for now

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/RestClient.java:799]: Sorry, the request returns the list of the job's signals after the 'add signal' operation

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java]: This file is not updated. It returns a set of string representing the result signals

 LogModel.getInstance().logMessage("Successfully add signal " + signal + " job " + jobId +
                                                  " has the following signals " + result);

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:277]: why do u pass menu ? u are not using it in the controller ?

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/SchedulerServiceImpl.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/SchedulerServiceImpl.java:1379]: return null; should be in the catch block

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:513]: why final ?

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:515]: event2 is never used

Comment (mboussaa) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:515]: if it is really needed, let's change event2 to just event

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (mboussaa) [inline scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:733]: why we have all path of the exception here? org.ow2.proactive_grid_cloud_portal.common.client.json.JSONException

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:302]: after sending the signal, the result will contain the list of actual signals
example: after I send the 'stop' signal, the result will be {kill} (the remaining signal)

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java:515]: I will rename it

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java]: I refactored it as class variable

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/view/grid/jobs/JobsListGrid.java]: I removed it

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (oanaBiancaSchiopu) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java]: I changed it and used the import

File: scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java
Comment (mboussaa) [reply scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/client/controller/JobsController.java:301]: could u give an example of what would this message dispaly:
"Successfully add signal " + signal + " job " + jobId +" has the following signals " + result
for me it is hard to read at the end

**Review by mboussaa — APPROVED** [review]
LGTM + Please, apply spotlessApply and optimize imports on all classes.
Thanks for taking care of the comments and nice work :))

**Comment by mboussaa** [comment]
retest this please!

**Comment by mboussaa** [comment]
retest this please!

**Pull request merged**
